src/node_jsvmapi.cc

```diff
@@ -1319,122 +1318,143 @@ napi_status napi_throw_type_error(napi_env e, const char* msg) {
 }
 
-napi_status napi_get_string_from_value(napi_env e,
-                                       napi_value v,
-                                       char* buf,
-                                       const int buf_size,
-                                       int* result) {
+napi_status napi_get_value_int32(napi_env e, napi_value v, int32_t* result) {
   NAPI_PREAMBLE(e);
   CHECK_ARG(result);
 
-  napi_valuetype v_type;
-  napi_status err = napi_get_type_of_value(e, v, &v_type);
-
-  // Consider: Should we pre-emptively set the length to zero ?
-  RETURN_STATUS_IF_FALSE(err == napi_ok, err);
+  v8::Local<v8::Value> value = v8impl::V8LocalValueFromJsValue(v);
+  if (!value->IsNumber()) return napi_set_last_error(napi_number_expected);
 
-  if (v_type == napi_number) {
-    v8::String::Utf8Value str(v8impl::V8LocalValueFromJsValue(v));
-    int len = str.length();
-    if (buf_size > len) {
-      memcpy(buf, *str, len);
-      *result = 0;
-    } else {
-      memcpy(buf, *str, buf_size - 1);
-      *result = len - buf_size + 1;
-    }
-  } else {
-    int len = 0;
-    err = napi_get_string_utf8_length(e, v, &len);
-    RETURN_STATUS_IF_FALSE(err == napi_ok, err);
-
-    int copied = v8impl::V8LocalValueFromJsValue(v).As<v8::String>()
-        ->WriteUtf8(
-            buf,
-            buf_size,
-            0,
-            v8::String::REPLACE_INVALID_UTF8 |
-            v8::String::PRESERVE_ONE_BYTE_NULL);
-    // add one for null ending
-    *result = len - copied + 1;
-  }
+  *result = value.As<v8::Int32>()->Value();
 
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_get_value_int32(napi_env e, napi_value v, int32_t* result) {
+napi_status napi_get_value_uint32(napi_env e, napi_value v, uint32_t* result) {
   NAPI_PREAMBLE(e);
   CHECK_ARG(result);
 
-  *result = v8impl::V8LocalValueFromJsValue(v)->Int32Value();
+  v8::Local<v8::Value> value = v8impl::V8LocalValueFromJsValue(v);
+  if (!value->IsNumber()) return napi_set_last_error(napi_number_expected);
+
+  *result = value.As<v8::Uint32>()->Value();
 
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_get_value_uint32(napi_env e, napi_value v, uint32_t* result) {
+napi_status napi_get_value_int64(napi_env e, napi_value v, int64_t* result) {
   NAPI_PREAMBLE(e);
   CHECK_ARG(result);
 
-  *result = v8impl::V8LocalValueFromJsValue(v)->Uint32Value();
+  v8::Local<v8::Value> value = v8impl::V8LocalValueFromJsValue(v);
+  if (!value->IsNumber()) return napi_set_last_error(napi_number_expected);
+
+  *result = value.As<v8::Integer>()->Value();
 
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_get_value_int64(napi_env e, napi_value v, int64_t* result) {
+napi_status napi_get_value_bool(napi_env e, napi_value v, bool* result) {
   NAPI_PREAMBLE(e);
   CHECK_ARG(result);
 
-  *result = v8impl::V8LocalValueFromJsValue(v)->IntegerValue();
+  v8::Local<v8::Value> value = v8impl::V8LocalValueFromJsValue(v);
+  if (!value->IsBoolean()) return napi_set_last_error(napi_boolean_expected);
```

Comment:
These napi_get_value_* API don't do coercion anymore, so what's the pattern we think integrators should use?

```c++
bool r;
napi_status s = napi_get_value_bool(env, possibly_bool, &r);
if (s == napi_boolean_expected) {
  // ?
} else if (s == napi_ok) {
  return r;
} else {
  CHECK_STATUS(napi_throw_error(env, "Boolean expected!"));
}
```

Or do we provide a ```napi_coerce_to_bool``` function?

```c++
bool r;
napi_value definitely_bool;
CHECK_STATUS(napi_coerce_to_bool(env, possibly_bool, &definitely_bool));
CHECK_STATUS(napi_get_value_bool(env, definitely_bool, &r);
return r;
```

Or push the logic into module?
```c++
bool r;
napi_valuetype t;
CHECK_STATUS(napi_get_type_of_value(env, possibly_bool, &t));
switch (t) {
case napi_boolean:
  CHECK_STATUS(napi_get_value_bool(env, possibly_bool, &r);
  return r;
case napi_number:
  int32_t temp;
  CHECK_STATUS(napi_get_value_int32(env, possibly_bool, &temp));
  return (bool)temp;
...
}
```